### DISCUSSION TITLE
Classless Routing 도입 이유

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#206] [Network]
.

**Comment by le2sky** [comment]
Classless Routing 프로토콜은 서브넷을 적절히 사용하여 IP 주소 공간을 더욱 효율적으로 활용합니다.
또한, 해당 프로토콜의 핵심 요소인 CIDR를 사용하면 같은 네트워크 주소 범위를 여러 작은 항목으로 나누는 대신 하나의 큰 항목으로 묶어서 사용할 수 있기 때문에 다음과 같은 이점이 있습니다.

- 라우팅 테이블의 크기를 줄일 수 있습니다.
- 라우터의 메모리 사용량을 낮출 수 있습니다.
- 라우팅 처리 속도를 향상시킬 수 있습니다.

**Comment by J-I-H-O** [comment]
**<IP 구조>**

IP 주소는 (네트워크 주소 + 호스트 주소)로 구성됩니다.

- **네트워크 주소 :** 호스트 들을 모은 네트워크를 칭합니다. 하나의 로컬 네트워크에 속한 호스트들의 네트워크 주소는 같아야 합니다.
- **호스트 주소 :** 하나의 네트워크에 존재하는 호스트를 구분하기 위한 주소입니다.

---

**<Classful Routing 프로토콜>**

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/3e69be4e-2a3d-4ec1-aa39-431949c92293)

Class를 나누어 네트워크 주소와 호스트 주소를 구분하는 경계 기준을 정합니다. (전통적 방법)
Class는 하나의 네트워크에서 몇 개의 호스트 IP까지 할당할 수 있는가에 대한 개념입니다. 
각각의 네트워크가 필요로 하는 호스트 수에 맞추어 적절히 분배하기 위해 등장했습니다.

- **A클래스**
    - 첫 번째 옥텟(`.` 으로 구분되는 영역)은 네트워크 주소를 나타내고, 나머지 부분은 호스트 주소를 나타냄
    - 하나의 네트워크가 가질 수 있는 호스트 수가 가장 많은 클래스
    - IP 주소를 32자리 2진수로 표현했을 때, 맨 앞자리 수가 항상 0
        - `0nnn nnnn . hhhh hhhh . hhhh hhhh . hhhh hhhh`
    - 0.0.0.0 ~ 127.255.255.255 까지 표현 가능
        - 이때 127은 자기 자신을 뜻하는 `루프백` 이므로, 실제로는 126.x.x.x 까지만 가질 수 있음
    - `2^24 - 2` 개의 호스트 주소를 사용할 수 있음
        - 이때 2개를 제외하는 이유는, 모든 호스트 주소의 비트가 0인 경우는 네트워크 주소로 사용하고, 모든 호스트 주소의 비트가 1인 경우는 브로드캐스트 주소로 사용하기 때문
            - 예) 13.0.0.0 는 네트워크 주소를 표현하기 위해서, 13.255.255.255 는 브로드캐스트 주소로 사용되기 때문에 호스트 IP로 사용할 수 없음
- **B클래스**
    - 첫 두 개의 옥텟은 네트워크 주소를 나타내고, 나머지 부분은 호스트 주소를 나타냄
    - IP 주소를 32자리 2진수로 표현했을 때, 맨 앞자리 두 수가 항상 10
        - `10nn nnnn . nnnn nnnn . hhhh hhhh . hhhh hhhh`
    - 128.0.0.0 ~ 191.255.255.255 까지 표현 가능
    - `2^16 - 2` 개의 호스트 주소를 사용할 수 있음
- **C클래스**
    - 첫 세 개의 옥텟은 네트워크 주소를 나타내고, 나머지 부분은 호스트 주소를 나타냄
    - IP 주소를 32자리 2진수로 표현했을 때, 맨 앞자리 세 수가 항상 110
        - `110n nnnn . nnnn nnnn . nnnn nnnn . hhhh hhhh`
    - 192.0.0.0 ~ 223.255.255.255 까지 표현 가능
    - `2^8 -2` 개의 호스트 주소를 사용할 수 있음

---

**<Classless Rounting 프로토콜>**

Classless Routing은 인터넷이 상용화되며 폭발적으로 증가한 사용량에 따른 IP 주소 부족을 해결하기 위해 등장했습니다.
전통적인 Classful Routing 방법에서는 클래스별로 고정된 서브넷 마스크를 사용해 네트워크 주소와 호스트 주소를 구분했습니다. 이로인해 실제로 필요한 IP 주소보다 훨씬 많은 주소가 할당되어 낭비되는 경우가 많았습니다.

Classless Routing에서는 IP 주소 클래스에 상관없이 서브넷 마스크를 자유롭게 설정할 수 있습니다. 이를 통해 네트워크 주소와 호스트 주소를 더 유연하게 관리할 수 있습니다.

---

**<CIDR: 사이더>**

CIDR(Classless Inter-Domain Routing, 사이더)는 Clessless Routing을 실질적으로 구현한 방식 중 하나입니다. 도메인간의 라우팅에 사용되는 IP 주소를 Classful 체계를 쓰는 것보다 더욱 능동적으로 할당하는 방법을 말합니다.

CIDR을 사용하면 필요한 만큼의 IP 주소만 할당할 수 있어 주소 낭비를 줄일 수 있고, 자유롭게 서브넷을 나눌 수 있어 유연한 네트워크 설계가 가능해집니다.
(서브넷팅, 한 개의 네트워크를 서브넷 마스크를 이용해 여러 개의 서브넷으로 분할)

또 여러 IP 주소를 하나의 경로로 묶어 하나의 라우팅 테이블 항목에 넣는 것(슈퍼넷팅)이 가능합니다.
여러개의 작은 네트워크를 하나의 큰 네트워크로 묶음으로써 라우팅 테이블의 엔트리 수를 줄일 수 있습니다. 
이러한 이점으로 인해  라우터의 메모리 사용을 줄이고, 성능을 향상시킬 수 있습니다.

---

**<슈퍼넷팅 예시>**

다음과 같은 네 개의 클래스 C 네트워크가 있다고 가정 해보겠습니다.

- `192.168.0.0/24`
- `192.168.1.0/24`
- `192.168.2.0/24`
- `192.168.3.0/24`

각각의 네트워크는 `/24` 서브넷 마스크를 가지고 있습니다. 이를 슈퍼넷팅을 통해 하나의 주소 범위로 묶으면 다음과 같습니다. 

- `192.168.0.0/22`

여기서 `/22`는 네트워크 주소가 22비트이고 호스트 주소가 10비트임을 의미합니다.
이는 `192.168.0.0` 부터 `192.168.3.255` 까지의 주소 범위를 포함합니다.

---

**<번외: 서브넷 마스크 표기법>**

서브넷 마스크는 IP 주소에서  `192.168.0.1 서브넷 마스크:255.255.255.0`  또는 `192.168.0.1/24` 와 같이 표현됩니다.
`255.255.255.0` 을 2진수로 나타내면 `1111 1111 . 1111 1111 . 1111 1111 . 0000 0000` 입니다. 
이때 1로 채워진 부분이 네트워크 주소, 0으로 채워진 부분이 호스트 주소가 됩니다.

이 때 `/24` 는 서브넷 마스크에서 1로 설정된 비트 수를 나타낸 것 입니다.
이처럼 IP 주소와 함께 서브넷 마스크를 슬래시로 표시하는 방법을 CIDR 표기법이라고 합니다. 

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/1bf51201-de58-4b86-887f-b73e465dae73)

만약 AWS VPC를 구성할 때 `10.0.0.0/16` 이라는 CIDR 값이 있다면, 이는 앞의 16비트를 네트워크 주소(고정 범위)로 사용하겠다는 의미입니다. 
자연스럽게 남은 16 비트는 호스트 주소(가변 범위)를 나타내게 되므로, `2^16 - 5` 개의 호스트를 할당할 수 있게 됩니다.

> 보통의 경우 IP 주소 범위의 가장 처음과 끝은 네트워크 주소(`10.0.0.0`)와 브로드캐스트 주소(`10.255.255.255`)로써 이 2개의 IP 주소를 제외하고 할당할 수 있습니다.
단, AWS에서는 자체 클라우드에서 설정해 사용하고 있는 IP가 있기 때문에 총 5개를 제외해야 합니다.
> 

https://code-lab1.tistory.com/34

https://velog.io/@borab/IP-%EC%A3%BC%EC%86%8C%EC%99%80-Classful-Classless

https://hstory0208.tistory.com/entry/%EB%84%A4%ED%8A%B8%EC%9B%8C%ED%81%AC-Class%EB%9E%80-CIDR-%EB%9E%80-%EC%B0%A8%EC%9D%B4%EC%A0%90-%EB%B0%8F-%EA%B0%9C%EB%85%90%EC%9D%84-%EC%89%BD%EA%B2%8C-%EC%9D%B4%ED%95%B4%ED%95%B4%EB%B3%B4%EC%9E%90-feat%EC%84%9C%EB%B8%8C%EB%84%B7%ED%8C%85-%EC%8A%88%ED%8D%BC%EB%84%B7%ED%8C%85

https://inpa.tistory.com/entry/WEB-%F0%9F%8C%90-CIDR-%EC%9D%B4-%EB%AC%B4%EC%96%BC-%EB%A7%90%ED%95%98%EB%8A%94%EA%B1%B0%EC%95%BC-%E2%87%9B-%EA%B0%9C%EB%85%90-%EC%A0%95%EB%A6%AC-%EA%B3%84%EC%82%B0%EB%B2%95

**Reply by J-I-H-O** [reply]
TODO: 세 줄 요약
